daemon/containerd/image_list.go

```diff
@@ -337,6 +346,111 @@ func (i *ImageService) setupFilters(ctx context.Context, imageFilters filters.Ar
 	}, nil
 }
 
+// setupLabelFilter parses filter args for "label" and "label!" and returns a
+// filter func which will check if any image config from the given image has
+// labels that match given predicates.
+func setupLabelFilter(store content.Store, fltrs filters.Args) (func(image images.Image) bool, error) {
+	type labelCheck struct {
+		key        string
+		value      string
+		onlyExists bool
+		negate     bool
+	}
+
+	var checks []labelCheck
+	for _, fltrName := range []string{"label", "label!"} {
+		for _, l := range fltrs.Get(fltrName) {
+			k, v, found := strings.Cut(l, "=")
+			err := labels.Validate(k, v)
+			if err != nil {
+				return nil, err
+			}
+
+			negate := strings.HasSuffix(fltrName, "!")
+
+			// If filter value is key!=value then flip the above.
+			if strings.HasSuffix(k, "!") {
+				k = strings.TrimSuffix(k, "!")
+				negate = !negate
+			}
+
+			checks = append(checks, labelCheck{
+				key:        k,
+				value:      v,
+				onlyExists: !found,
+				negate:     negate,
+			})
+		}
+	}
+
+	return func(image images.Image) bool {
+		ctx := context.TODO()
+
+		errFoundConfig := errors.New("success, found matching config")
+		err := images.Dispatch(ctx, presentChildrenHandler(store, images.HandlerFunc(func(ctx context.Context, desc v1.Descriptor) (subdescs []v1.Descriptor, err error) {
+			if images.IsConfigType(desc.MediaType) {
+				// Subset of ocispec.Image that only contains Labels
+				var cfg struct {
+					Config struct {
+						Labels map[string]string `json:"Labels,omitempty"`
+					} `json:"Config,omitempty"`
+				}
+				err := readConfig(ctx, store, desc, &cfg)
+				if err != nil {
+					return nil, err
+				}
+
+				for _, check := range checks {
+					value, exists := cfg.Config.Labels[check.key]
+
+					if check.onlyExists {
+						// label! given without value, check if doesn't exist
+						if check.negate {
+							// Label exists, config doesn't match
+							if exists {
+								return nil, nil
+							}
+						} else {
+							// Label should exist
+							if !exists {
+								// Label doesn't exist, config doesn't match
+								return nil, nil
+							}
+
+						}
+						continue
+					} else if !exists {
+						// We are checking value and label doesn't exist.
+						return nil, nil
+					}
+
+					valueEquals := value == check.value
+					if valueEquals == check.negate {
+						return nil, nil
+					}
+				}
+
+				return nil, errFoundConfig
```

Comment:
Nope, we want to exit as soon as we find a matching config from that image - if we already know that this image has a config that matches filters, we don't need to check anything else. Returning an "operation failed successfully" will make the whole Dispatch stop and we'll know it stopped because it found what we were looking for.

Reply:
Oh! Gotcha, yes, the error is to make the `images.Dispatch` cancel (it doesn't if `images.HandlerFunc()` returns a nil-error)!

Got it now. Perhaps that would warrant a comment either at the `return nil, errFoundConfig` line, or at the `errFoundConfig := errors.New(....` line (or perhaps _both_); WDYT?

Reply:
Added small comments to both.